### ISSUE TITLE
Can't reactivate fvm after pub cache repair

### ISSUE THREAD

**Issue by RomanKapshuk** [leoafarias/fvm#352]
As a result of `fvm flutter pub cache repair` at the end of execution I get

```
Error: Couldn't resolve the package 'fvm' in 'package:fvm/src/runner.dart'.
../../.pub-cache/hosted/pub.dartlang.org/fvm-2.2.3/bin/main.dart:3:8: Error: Not found: 'package:fvm/src/runner.dart'
import 'package:fvm/src/runner.dart';
       ^
../../.pub-cache/hosted/pub.dartlang.org/fvm-2.2.3/bin/main.dart:6:14: Error: Method not found: 'FvmCommandRunner'.
  exit(await FvmCommandRunner().run(args));
             ^^^^^^^^^^^^^^^^
Failed to reactivate fvm 2.2.3: Failed to build fvm:main:
Error: Couldn't resolve the package 'fvm' in 'package:fvm/src/runner.dart'.
../../.pub-cache/hosted/pub.dartlang.org/fvm-2.2.3/bin/main.dart:3:8: Error: Not found: 'package:fvm/src/runner.dart'
import 'package:fvm/src/runner.dart';
       ^
../../.pub-cache/hosted/pub.dartlang.org/fvm-2.2.3/bin/main.dart:6:14: Error: Method not found: 'FvmCommandRunner'.
  exit(await FvmCommandRunner().run(args));
             ^^^^^^^^^^^^^^^^
Binstubs exist for non-activated packages:
  From fvm: fvm
Reactivated 1 package.
Failed to reactivate 1 package:
- fvm
pub finished with exit code 69
```

**Comment by leoafarias** [comment]
@RomanKapshuk sorry for the delay on this. Are you able to reactivate fvm by calling dart or flutter sdk directly?

**Comment by tazik561** [comment]
> @RomanKapshuk sorry for the delay on this. Are you able to reactivate fvm by calling dart or flutter sdk directly?

For first time fvm is known by terminal, means when I run ` fvm --version` I got true answer. But after running this command `flutter pub cache repair` and get `Error: Couldn't resolve the package 'fvm' in 'package:fvm/src/runner.dart'.` error, fvm seems broken and when I run ` ` fvm --version` I get `zsh: command not found: fvm` and I have to run `dart pub global activate fvm` again.

**Comment by leoafarias** [comment]
@tazik561 thanks for the details… will have to do some research on this, as I am not sure what fvm is doing differently than any other package.

**Comment by btrautmann** [comment]
FWIW, I've been noticing this with other packages, so I don't think this is specific to fvm. I've noticed the same with `melos` and a package that I've been working on and have installed locally.

**Comment by tazik561** [comment]
> rking on and h

I think is because of fvm, I changed FVM to official SDK and the problem is gone. Now I am using official SDK and I don't have a problem .

**Comment by leoafarias** [comment]
I believe I know what is the problem, and has something that has been brought up before. The process is being run by FVM, and during the repair, process fvm is removed.

I am looking into what some of the options might be for the `pub global activate fvm` install method. The issue will not happen if you install through brew or chocolatey.

Will investigate ways of detaching that process to see if it helps.

**Comment by leoafarias** [comment]
Ok so I took a look, and the solution is to check if the pub command being run might impact the fvm install. 
- Check if command might impact fvm install
- Check if fvm is installed through pub global (not brew, chocolatey)

If those are true it will go ahead and stop execution and output an explanation and the command you should add to your terminal.

Also since in the case the installation is not repaired and just removed like in the case of `dart pub cache clean`, it will also output a command to the direct executable to reinstall fvm.

Here is a screenshot:

![CleanShot 2021-11-12 at 10 59 23](https://user-images.githubusercontent.com/435833/141496473-a15622f0-84b4-4224-b0ad-372c65446c3f.png)

**Comment by btrautmann** [comment]
FWIW I run into this when running `dart pub cache repair` directly, not via `fvm` -- not sure if that matters

**Comment by ccastroelo** [comment]
Sorry I don't know if I understand correctly.

In a power shell as an administrator. I ran: 
dart pub global deactivate fvm
and then: 
dart pub global activate fvm

I closed the window and opened it again as an administrator and ran:
fvm dart pub cache repair

I received the following:
Building package executables...
Failed to build fvm:main:
Error: Couldn't resolve the package 'fvm' in 'package:fvm/src/runner.dart'.
AppData/Roaming/Pub/Cache/hosted/pub.dartlang.org/fvm-2.2.4/bin/main.dart:3:8: Error: Not found: 'package:fvm/src/runner.dart'
import 'package:fvm/src/runner.dart';
       ^
AppData/Roaming/Pub/Cache/hosted/pub.dartlang.org/fvm-2.2.4/bin/main.dart:6:14: Error: Method not found: 'FvmCommandRunner'.
  exit(await FvmCommandRunner().run(args));
             ^^^^^^^^^^^^^^^^
Failed to reactivate fvm 2.2.4: Failed to build fvm:main:
Error: Couldn't resolve the package 'fvm' in 'package:fvm/src/runner.dart'.
AppData/Roaming/Pub/Cache/hosted/pub.dartlang.org/fvm-2.2.4/bin/main.dart:3:8: Error: Not found: 'package:fvm/src/runner.dart'
import 'package:fvm/src/runner.dart';
       ^
AppData/Roaming/Pub/Cache/hosted/pub.dartlang.org/fvm-2.2.4/bin/main.dart:6:14: Error: Method not found: 'FvmCommandRunner'.
  exit(await FvmCommandRunner().run(args));
             ^^^^^^^^^^^^^^^^
Binstubs exist for non-activated packages:
  From fvm: fvm
Reactivated 1 package.
Failed to reactivate 1 package:
- fvm

**Comment by leoafarias** [comment]
As I mentioned pub cache repair is essentially removing the FVM’s own dependencies. That is why you are getting this error. The PR attached once releases will warns the user that the install will be modified and provides instructions on how to execute it safely

**Comment by ccastroelo** [comment]
>> As I mentioned pub cache repair is essentially removing the FVM’s own dependencies. That is why you are getting this error. 

ok, I understand.

>> The PR attached once releases will warns the user that the install will be modified and provides instructions on how to execute it safely.

OK, but in all the commands that execute in none the instructions to solve the problem appear. Which command should I run for this?

Léo, 1. Muito bom o FVM, quebra um galhão. 2. desculpa, mas to boiando nesse problema.  Pois é, em nenhum comando q dei apareceu informando que a instalação será modificada. Qual é o comando devo rodar? (PR é pub cache repair? se for, eu rodo ele, mas ele só me retorna o erro, como listado acima, sem as  instruções de como executar ele safety.)

**Comment by leoafarias** [comment]
@ccastroelo this update has not been pushed live add me on telegram and we can discuss. @leoafarias

**Comment by ccastroelo** [comment]
> @ccastroelo this update has not been pushed live add me on telegram and we can discuss. @leoafarias

please add me. I have never used the telegram and for some reason my account is blocked for non-contacts. @ccastroelo

**Comment by ccastroelo** [comment]
After a few unsuccessful attempts I chose to delete the flutter and fvm folders and reinstall everything. And from then on I only got the rights issue when I road the fvm flutter pub get in the project.

As you recommended, I deleted the project's .fvm folder, ran "use" again and everything worked. I believe that with the previous instructions the problem would have been solved, but as there was the old configuration in the project's .fvm and I did the mistake of not deleting it, I was stuck.

Leo, thank you so much for your help and patience.
